serial.go

```diff
@@ -146,60 +147,118 @@ func (sp *SerialPortList) List() {
 	}
 }
 
-func (sp *SerialPortList) Update() {
-	if !sp.enumerationLock.TryLock() {
-		// already enumerating...
-		return
+// Run is the main loop for port discovery and management
+func (sp *SerialPortList) Run() {
+	for retries := 0; retries < 10; retries++ {
+		sp.runSerialDiscovery()
+
+		logrus.Errorf("Serial discovery stopped working, restarting it in 10 seconds...")
+		time.Sleep(10 * time.Second)
 	}
-	defer sp.enumerationLock.Unlock()
+	logrus.Errorf("Failed restarting serial discovery. Giving up...")
+}
 
-	livePorts, _ := enumerator.GetDetailedPortsList()
-	// TODO: report error?
+func (sp *SerialPortList) runSerialDiscovery() {
+	// First ensure that all the discoveries are available
+	if err := Tools.Download("builtin", "serial-discovery", "latest", "keep"); err != nil {
+		logrus.Errorf("Error downloading serial-discovery: %s", err)
+		panic(err)
+	}
+	sd, err := Tools.GetLocation("serial-discovery")
+	if err != nil {
+		logrus.Errorf("Error downloading serial-discovery: %s", err)
+		panic(err)
+	}
+	d := discovery.NewClient("serial", sd+"/serial-discovery")
+	dLogger := logrus.WithField("discovery", "serial")
+	if *verbose {
+		d.SetLogger(dLogger)
+	}
+	d.SetUserAgent("arduino-create-agent/" + version)
+	if err := d.Run(); err != nil {
+		logrus.Errorf("Error running serial-discovery: %s", err)
+		panic(err)
+	}
+	defer d.Quit()
 
-	var ports []SpPortItem
-	for _, livePort := range livePorts {
-		if !livePort.IsUSB {
-			continue
-		}
-		vid, pid := "0x"+livePort.VID, "0x"+livePort.PID
-		if vid == "0x0000" || pid == "0x0000" {
-			continue
-		}
-		if portsFilter != nil && !portsFilter.MatchString(livePort.Name) {
-			logrus.Debugf("ignoring port not matching filter. port: %v\n", livePort)
-			continue
-		}
+	events, err := d.StartSync(10)
```

Comment:
Just out of curiosity, why exactly 10? Should be high enough the guarantee us to cover possible initial events?

Reply:
Even 1 should be good enough for 99% of cases, I've put 10 for good measure, it's almost free.